File: BOJ/Graph/9466.cpp
// BOJ 9466. 텀 프로젝트
// dfs로 탐색하면서 사이클이 있는 경우, 사이클의 요소 개수를 카운트.

#include <bits/stdc++.h>

using namespace std;


int T, n, cnt;
int visited[100001];
int finished[100001];
int s[100001];

void dfs(int x) {
    visited[x] = 1;

    int nxt = s[x];
    if (!visited[nxt]) {
        dfs(nxt);
    } else {
        if (!finished[nxt]) {
            for (int i=nxt; i!=x; i=s[i]) {
                cnt++;
            }
            cnt++;
        }
    }
    finished[x] = 1;
}

int main() {
    cin >> T;

    while (T--) {
        cin >> n;
        for (int i=1; i<=n; i++) {
            cin >> s[i];
        }

        memset(visited, 0, sizeof(visited));
        memset(finished, 0, sizeof(finished));

        cnt = 0;
        for (int i=1; i<=n; i++) {
            if (!visited[i]) {
                dfs(i);
            }
        }

        cout << n-cnt << endl;
    }

}

File: BOJ/Graph/9466.py
# BOJ 9466. 텀 프로젝트
# dfs로 탐색하며 route를 기록.
# 만약 사이클이 있는 경우, route에서 사이클 부분의 길이를 구함.

import sys
input = sys.stdin.readline
sys.setrecursionlimit(1000000)
global s, cycleCnt
global visited, parent

def dfs(x):
    global cycleCnt
    visited[x] = 1
    route.append(x)
    nxt = s[x]
    if not visited[nxt]:
        dfs(nxt)
    else:
        if nxt in route:
            cycleCnt += len(route[route.index(nxt):])
            return

if __name__=='__main__':
    T = int(input())
    for _ in range(T):
        n = int(input())
        s = [0]+list(map(int, input().split()))
        visited = [0]*(n+1)
        parent = [0]*(n+1)
        cycleCnt = 0
        
        for i in range(1, n+1):
            if not visited[i]:
                route = []
                dfs(i)

        print(n-cycleCnt)
